server/src/main/java/org/opensearch/transport/ClusterConnectionManager.java

```diff
@@ -122,12 +131,19 @@ public void connectToNode(
         ConnectionValidator connectionValidator,
         ActionListener<Void> listener
     ) throws ConnectTransportException {
+        logger.trace("connecting to node [{}]", node);
         ConnectionProfile resolvedProfile = ConnectionProfile.resolveConnectionProfile(connectionProfile, defaultProfile);
         if (node == null) {
             listener.onFailure(new ConnectTransportException(null, "can't connect to a null node"));
             return;
         }
 
+        // if node-left is still in progress, we fail the connect request early
+        if (pendingDisconnections.contains(node)) {
+            listener.onFailure(new IllegalStateException("cannot make a new connection as disconnect to node [" + node + "] is pending"));
+            return;
+        }
+
```

Comment:
connect requests fail - saw this happening as part of join request flow.

join request keeps retrying infinitely. it would succeed after the node-left task has finished committing state on cluster-manager

Reply:
Can we optimise this to maintain not more than two connection per node one old and one new. We cut over from old to new connection on new node join. This should allow faster re-connections while waiting on older connections to drain?

Reply:
closed on this, we are good with the current approach itself